go-controller/pkg/ovn/network_controller_info.go

```diff
@@ -0,0 +1,612 @@
+	lbCache, err := ovnlb.GetLBCache(nci.nbClient)
+	if err != nil {
+		return fmt.Errorf("failed to get load_balancer cache for node %s: %v", nodeName, err)
+	}
+	lbCache.RemoveSwitch(switchName)
+
+	// Remove the logical switch associated with the node
+	err = libovsdbops.DeleteLogicalSwitch(nci.nbClient, switchName)
+	if err != nil {
+		return fmt.Errorf("failed to delete logical switch %s: %v", switchName, err)
+	}
+
+	logicalRouterName := types.OVNClusterRouter
+	logicalRouter := nbdb.LogicalRouter{Name: logicalRouterName}
+	logicalRouterPort := nbdb.LogicalRouterPort{
+		Name: types.RouterToSwitchPrefix + switchName,
+	}
+	err = libovsdbops.DeleteLogicalRouterPorts(nci.nbClient, &logicalRouter, &logicalRouterPort)
+	if err != nil {
+		return fmt.Errorf("failed to delete router port %s: %v", logicalRouterPort.Name, err)
+	}
+
+	return nil
+}
+
+// updates the list of nodes if the given node manages its hostSubnets; returns its hostSubnets if any
+func (nci *NetworkControllerInfo) updateNodesManageHostSubnets(node *kapi.Node,
+	masterSubnetAllocator *subnetallocator.HostSubnetAllocator, foundNodes sets.String) []*net.IPNet {
+	if noHostSubnet(node) {
+		return []*net.IPNet{}
+	}
+	hostSubnets, _ := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+	foundNodes.Insert(node.Name)
+
+	klog.V(5).Infof("Node %s contains subnets: %v", node.Name, hostSubnets)
+	if err := masterSubnetAllocator.MarkSubnetsAllocated(node.Name, hostSubnets...); err != nil {
+		utilruntime.HandleError(err)
+	}
+	return hostSubnets
+}
+
+func (nci *NetworkControllerInfo) requestAddPodOnNode(nodeName string) []error {
+	errs := []error{}
+	options := metav1.ListOptions{
+		FieldSelector:   fields.OneTermEqualSelector("spec.nodeName", nodeName).String(),
+		ResourceVersion: "0",
+	}
+	pods, err := nci.client.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), options)
+	if err != nil {
+		errs = append(errs, err)
+		klog.Errorf("Unable to list existing pods on node: %s, existing pods on this node may not function",
+			nodeName)
+	} else {
+		klog.V(5).Infof("When adding node %s, found %d pods to add to retryPods", nodeName, len(pods.Items))
+		for _, pod := range pods.Items {
+			pod := pod
+			if util.PodCompleted(&pod) {
+				continue
+			}
+			klog.V(5).Infof("Adding pod %s/%s to retryPods", pod.Namespace, pod.Name)
+			err = nci.retryPods.AddRetryObjWithAddNoBackoff(&pod)
+			if err != nil {
+				errs = append(errs, err)
+				klog.Errorf("Failed to add pod %s/%s to retryPods: %v", pod.Namespace, pod.Name, err)
+			}
+		}
+	}
+	nci.retryPods.RequestRetryObjs()
+	return errs
+}
+
+func (nci *NetworkControllerInfo) updateL3TopologyVersion() error {
+	currentTopologyVersion := strconv.Itoa(types.OvnCurrentTopologyVersion)
+	clusterRouterName := types.OVNClusterRouter
+	logicalRouter := nbdb.LogicalRouter{
+		Name:        clusterRouterName,
+		ExternalIDs: map[string]string{"k8s-ovn-topo-version": currentTopologyVersion},
```

Comment:
Do we plan to have different topology versions per topology? If not, wouldn't the topology version be better managed from a more centralized place?

Reply:
again, I'd like to get each controller to be independent, so each controller knows exactly if its own topology is upgraded to the right version and updated its own version number, regardless if they share the same topology version or not.

one thing that tbd is possibly the topology version needs to know by the node. today, that was done by the default controller. Not sure if one day, there is a need to only update the version when all the controllers are done updating its topology version. But since it is now not a requirement, I didn't change the current behavior.

Reply:
Ok. I think this is a good topic to discuss in the meeting.

Reply:
I think in the future it could be highly likely that one network type has a topology version change that the other network type does not care about. For example if we had secondary networks in the past, and we went from single join switch to multiple or vice versa, only one of those would care about topology version change.

Reply:
But if it doesn't care it might as well update to the last topology version anyway.
I guess the question is if we have an actual reason to have multiple controller types at different versions.